heat/core/manipulations.py

```diff
@@ -2074,115 +2083,188 @@ def sort(a: DNDarray, axis: int = -1, descending: bool = False, out: Optional[DN
+    target_cumsum = np.cumsum(counts)
+    for idx in np.ndindex(local_sorted.shape[1:]):
+        idx_slice = [slice(None)] + [slice(ind, ind + 1) for ind in idx]
+        current_counts = partition_matrix[idx_slice].reshape(-1).tolist()
+        current_cumsum = list(np.cumsum(current_counts))
+        for proc in range(size):
+            # process has to many values which will be sent to higher ranks
+            if current_cumsum[proc] > target_cumsum[proc]:
+                first = next(i for i in range(size) if send_vec[idx][:, i].sum() < counts[i])
+                last = next(
+                    i
+                    for i in range(size + 1)
+                    if i == size or current_cumsum[proc] < target_cumsum[i]
+                )
+                sent = 0
+                for i, x in enumerate(counts[first:last]):
+                    # Each following process gets as many elements as it needs
+                    amount = int(x - send_vec[idx][:, first + i].sum())
+                    send_vec[idx][proc][first + i] = amount
+                    current_counts[first + i] += amount
+                    sent = send_vec[idx][proc][: first + i + 1].sum().item()
+                if last < size:
+                    # Send all left over values to the highest last process
+                    amount = partition_matrix[proc][idx]
+                    send_vec[idx][proc][last] = int(amount - sent)
+                    current_counts[last] += int(amount - sent)
+            elif current_cumsum[proc] < target_cumsum[proc]:
+                # process needs values from higher rank
+                first = (
+                    0
+                    if proc == 0
+                    else next(
+                        i for i, x in enumerate(current_cumsum) if target_cumsum[proc - 1] < x
                     )
-                current_counts[proc] = counts[proc]
-                current_cumsum = list(np.cumsum(current_counts))
+                )
+                last = next(i for i, x in enumerate(current_cumsum) if target_cumsum[proc] <= x)
+                for i, x in enumerate(partition_matrix[idx_slice][first:last]):
+                    # Taking as many elements as possible from each following process
+                    send_vec[idx][first + i][proc] = int(x - send_vec[idx][first + i].sum())
+                    current_counts[first + i] = 0
+                # Taking just enough elements from the last element to fill the current processes tensor
+                send_vec[idx][last][proc] = int(target_cumsum[proc] - current_cumsum[last - 1])
+                current_counts[last] -= int(target_cumsum[proc] - current_cumsum[last - 1])
+            else:
+                # process doesn't need more values
+                send_vec[idx][proc][proc] = partition_matrix[proc][idx] - send_vec[idx][proc].sum()
+            current_counts[proc] = counts[proc]
+            current_cumsum = list(np.cumsum(current_counts))
```

Comment:
numpy again. im not sure if it makes a large difference to put it in torch though. that can be said about most of these things to be fair, but if something uses it on a GPU then it needs to be transferred there

Reply:
legacy code, fixed